src/nupic/algorithms/AnomalyLikelihood.cpp

```diff
@@ -0,0 +1,505 @@
+
+using namespace std;
+using namespace nupic;
+using namespace nupic::util;
+using namespace nupic::algorithms::anomaly;
+
+namespace nupic {
+  namespace algorithms {
+    namespace anomaly {
+
+
+  /**
+      NOTE: Anomaly likelihood scores are reported at a flat 0.5 for
+    learningPeriod + estimationSamples iterations.
+
+    claLearningPeriod and learningPeriod are specifying the same variable,
+    although claLearningPeriod is a deprecated name for it.
+
+    @param learningPeriod (claLearningPeriod: deprecated) - (int) the number of
+      iterations required for the algorithm to learn the basic patterns in the
+      dataset and for the anomaly score to 'settle down'. The default is based
+      on empirical observations but in reality this could be larger for more
+      complex domains. The downside if this is too large is that real anomalies
+      might get ignored and not flagged.
+
+    @param estimationSamples - (int) the number of reasonable anomaly scores
+      required for the initial estimate of the Gaussian. The default of 100
+      records is reasonable - we just need sufficient samples to get a decent
+      estimate for the Gaussian. It's unlikely you will need to tune this since
+      the Gaussian is re-estimated every 10 iterations by default.
+
+    @param historicWindowSize - (int) size of sliding window of historical
+      data points to maintain for periodic reestimation of the Gaussian. Note:
+      the default of 8640 is based on a month's worth of history at 5-minute
+      intervals.
+
+    @param reestimationPeriod - (int) how often we re-estimate the Gaussian
+      distribution. The ideal is to re-estimate every iteration but this is a
+      performance hit. In general the system is not very sensitive to this
+      number as long as it is small relative to the total number of records
+      processed.
+
+  **/
+AnomalyLikelihood::AnomalyLikelihood(UInt learningPeriod, UInt estimationSamples, UInt historicWindowSize, UInt reestimationPeriod, UInt aggregationWindow) :
+    learningPeriod(learningPeriod),
+    reestimationPeriod(reestimationPeriod),
+    averagedAnomaly(aggregationWindow) {
+        iteration = 0;
+        probationaryPeriod = learningPeriod+estimationSamples;
+        assert(historicWindowSize >= estimationSamples); // cerr << "estimationSamples exceeds historicWindowSize";
+        assert(aggregationWindow < reestimationPeriod && reestimationPeriod < historicWindowSize);
+        
+        runningAverageAnomalies.set_capacity(historicWindowSize); 
+        runningLikelihoods.set_capacity(historicWindowSize);
+        runningRawAnomalyScores.set_capacity(historicWindowSize);
+        runningRawValues.set_capacity(historicWindowSize);
+        assert(runningLikelihoods.capacity() == historicWindowSize);
+    }
+
+    
+Real AnomalyLikelihood::anomalyProbability(Real rawValue, Real anomalyScore, int timestamp) { //TODO "rawValue" could be typed to T rawValue; //TODO is rawValue ever used?
```

Comment:
You're correct that it's weird that `rawValue` is a specific type. Really it's weird that this class even receives the raw values. It's because of the big flatline hack. I think ultimately that hack should be removed from the anomaly likelihood code and the caller should handle it. But that would involve changing nupic. Maybe the best thing is to remove the hack in the C++ code and just document that this code doesn't offer that hack "feature".